staging/src/k8s.io/api/core/v1/types.go

```diff
@@ -1765,6 +1765,7 @@ type ContainerPort struct {
 	// Protocol for port. Must be UDP, TCP, or SCTP.
 	// Defaults to "TCP".
 	// +optional
+	// +kubebuilder:default=TCP
```

Comment:
I'm not opposed to adding indicators of default values to kubernetes types, but I would not expect kubebuilder annotations on core types

Reply:
+1 to what you say here @liggitt, although there is a real problem right now in that:

1) the default value *must* be specified when something is used as a list map key
2) there is no way for kubebuilder to introspect and understand what default value is needed (in the case where it _can_ be defined declaratively)

I notice in https://github.com/kubernetes/kubernetes/issues/91395#issuecomment-634029609 you mentioned a KEP to try and resolve this. This may be something I can pick up - presumably this is mostly defining the required/expected annotation (for the simple case where the default _is_ declarative), and potentially "leaving room" for us to express "this default is not simple/declarative".

Is that a correct analysis of the current state? Are you aware of anyone else trying to get this document/plan together?

Reply:
For some additional context on why this is important to solve, the current "workaround" for this that I have seen either involves either:

1) users running `sed` to fix up their CRDs
2) users forking the k8s.io/api repo (and effectively just applying this patch)
3) users running `sed` to mark the affected fields as `required`

Neither of these is really ideal, and it'd be great to 'nip this' before these sorts of patterns proliferate across the ecosystem

Reply:
In the meantime, kubebuilder annotations were added to https://github.com/kubernetes/kubernetes/pull/92660

Those are a bit different 1) because we expect those condition types to be reused in many custom resources, and 2) they are all about adding validation, not defining defaulting.

The main things I expect from a "default" annotation:
* it must be valid JSON (kube-builder appears to allow unquoted keys)
* it must be declarative for properties used as keys in a listType=map, because server-side apply will start using those inspected values

As a strawman, `+default=<valid JSON that can be unmarshaled into the field>` seems reasonable